js/web/lib/onnxjs/backends/webgl/ops/reshape-packed.ts

```diff
@@ -1,82 +1,60 @@
-    let inputLayout: TextureLayout;
-    inputLayout = handler.getOrCreateTextureLayout(inputs[0], 4, true, originInputShape, true);
-    if (originInputShape.length !== 3) {
-      const originalInputLayout = inputLayout;
-      // if originShape is not a 3D shape, create texture layout from the processed shape.
-      inputLayout = handler.createTextureLayoutFromShape(
-          this.inputShape3D, 4, this.inputShape3D, {isPacked: true, reverseWH: true});
-      // if the processed input shape produces texture layout differnt from the original
-      // one, the run data has to use the processed (3D) input shape later.
-      this.needSqueezeInputData =
-          (inputLayout.height !== originalInputLayout.height) || (inputLayout.width !== originalInputLayout.width);
-    }
+import {ProgramInfo, ProgramInfoLoader, ProgramMetadata, TextureType} from '../types';
 
-    this.outputShape = ShapeUtil.calculateReshapedDims(originInputShape, inputs[1].integerData);
-    const squeezedOutputShape = processDims3D(this.outputShape);
-
-    this.outputLayout = handler.createTextureLayoutFromShape(
-        squeezedOutputShape, 4, squeezedOutputShape, {isPacked: true, reverseWH: true});
-
-    let mainLoop = '';
-    for (let i = 0; i < 4; i++) {
-      let outputCoords = '';
-      switch (i) {
-        case 0:
-          outputCoords = 'outputCoords = rc;';
-          break;
-        case 1:
-          outputCoords = 'outputCoords = ivec3(rc.x, rc.y+1, rc.z);';
-          break;
-        case 2:
-          outputCoords = 'outputCoords = ivec3(rc.x, rc.y, rc.z+1);';
-          break;
-        case 3:
-          outputCoords = 'outputCoords = ivec3(rc.x, rc.y+1, rc.z+1);';
-          break;
-        default:
-          throw new Error();
-      }
+import {unpackFromChannel} from './packing-utils';
 
-      mainLoop += `
+const createPackedReshape3DProgramMetadata = (outputShape3D: readonly number[]) =>
+    ({name: 'Reshape (packed)', inputTypes: [TextureType.packed], inputNames: ['A'], cacheHint: `${outputShape3D}`});
+
+const createPackedReshape3DProgramInfo =
+    (handler: WebGLInferenceHandler, input3D: Tensor, metadata: ProgramMetadata, outputShape3D: readonly number[]):
+        ProgramInfo => {
+          // For packed reshape, we need to re-arrange texel data for output shape.
+          // Our pack is designed to pack a 2x2 tile in last h and w dimension, so
+          // for the reshaped new tensor, we just need to re-arrange the last h and
+          // w dimension. For any shape that is not in 3D, i.e. [batch, W, H], we
+          // first convert it to 3D by collapsing other dimension to batch dim, then
+          // process with the last two dimensions.
+          // Note: we only need the shape tensor to calculate output shape, so the
+          // content in shape tensor is never uploaded to GPU. It is always kept in CPU.
+          // TODO: optimize the algorithm -- in some cases, if the last two dims are
```

Comment:
this TODO has been done in the isReshapeCheap() function?

Reply:
removed the 2nd TODO and move the whole comments above `isReshapeCheap()`.